proxy/src/coco/source.rs

```diff
@@ -0,0 +1,452 @@
+    maybe_prefix: Option<String>,
+) -> Result<Tree, error::Error> {
+    let (mut repo, project) = client.get_repo(project_urn).await?;
+    let locked = repo.locked();
+    let mut browser = locked.browser()?;
+
+    let revision = maybe_revision.unwrap_or_else(|| project.default_branch().to_string());
+    let prefix = maybe_prefix.unwrap_or_default();
+
+    browser.revspec(&revision)?;
+
+    let path = if prefix == "/" || prefix == "" {
+        surf::file_system::Path::root()
+    } else {
+        surf::file_system::Path::from_str(&prefix)?
+    };
+
+    let root_dir = browser.get_directory()?;
+    let prefix_dir = if path.is_root() {
+        root_dir
+    } else {
+        root_dir
+            .find_directory(path.clone())
+            .ok_or_else(|| error::Error::PathNotFound)?
+    };
+    let mut prefix_contents = prefix_dir.list_directory();
+    prefix_contents.sort();
+
+    let entries_results: Result<Vec<TreeEntry>, error::Error> = prefix_contents
+        .iter()
+        .map(|(label, system_type)| {
+            let entry_path = if path.is_root() {
+                surf::file_system::Path::new(label.clone())
+            } else {
+                let mut p = path.clone();
+                p.push(label.clone());
+                p
+            };
+            let mut commit_path = surf::file_system::Path::root();
+            commit_path.append(entry_path.clone());
+
+            let info = Info {
+                name: label.to_string(),
+                object_type: match system_type {
+                    surf::file_system::SystemType::Directory => ObjectType::Tree,
+                    surf::file_system::SystemType::File => ObjectType::Blob,
+                },
+                last_commit: None,
+            };
+
+            Ok(TreeEntry {
+                info,
+                path: entry_path.to_string(),
+            })
+        })
+        .collect();
+
+    let mut entries = entries_results?;
+
+    // We want to ensure that in the response Tree entries come first. `Ord` being derived on
+    // the enum ensures Variant declaration order.
+    //
+    // https://doc.rust-lang.org/std/cmp/trait.Ord.html#derivable
+    entries.sort_by(|a, b| a.info.object_type.cmp(&b.info.object_type));
+
+    let last_commit = if path.is_root() {
+        Some(Commit::from(browser.get().first()))
+    } else {
+        None
+    };
+    let name = if path.is_root() {
+        "".into()
+    } else {
+        let (_first, last) = path.split_last();
+        last.to_string()
+    };
+    let info = Info {
+        name,
+        object_type: ObjectType::Tree,
+        last_commit,
+    };
+
+    Ok(Tree {
+        path: prefix,
+        entries,
+        info,
+    })
+}
+
+/// Initialize a [`radicle_surf::git::git2::Repository`] at the given path.
+///
+/// # Errors
+///
+/// Will return [`error::Error`] if any of the git interactions break.
+pub fn init_repo(path: String) -> Result<(), error::Error> {
```

Comment:
It'd be good track down where this is used. Using the link backend to this would make more sense possibly?